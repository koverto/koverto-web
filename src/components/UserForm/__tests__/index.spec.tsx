import { render, wait, waitForElement } from "@testing-library/react"
import * as React from "react"
import { UserForm, UserFormFields } from ".."

const onSubmit = jest.fn()

const Component = ({ loading = false }: { loading?: boolean }): JSX.Element => (
  <UserForm
    fields={[
      UserFormFields.EMAIL,
      UserFormFields.NAME,
      UserFormFields.PASSWORD,
    ]}
    label="User"
    loading={loading}
    onSubmit={onSubmit}
  />
)

describe("UserForm", () => {
  it("matches the snapshot", () => {
    const { container } = render(<Component />)
    expect(container).toMatchSnapshot()
  })

  it("requires the email field", async () => {
    const { getByTestId, queryByTestId } = render(<Component />)
    const button = await waitForElement(() => getByTestId("submit"))

    button.click()

    await wait(() => expect(queryByTestId("email-error")).toBeTruthy())
    expect(onSubmit).not.toBeCalled()
  })

  it("requires the name field", async () => {
    const { getByTestId, queryByTestId } = render(<Component />)
    const button = await waitForElement(() => getByTestId("submit"))

    button.click()

    await wait(() => expect(queryByTestId("name-error")).toBeTruthy())
    expect(onSubmit).not.toBeCalled()
  })

  it("requires the password field", async () => {
    const { getByTestId, queryByTestId } = render(<Component />)
    const button = await waitForElement(() => getByTestId("submit"))

    button.click()

    await wait(() => expect(queryByTestId("password-error")).toBeTruthy())
    expect(onSubmit).not.toBeCalled()
  })

  it("calls the onSubmit callback", async () => {
    const { getByTestId } = render(<Component />)
    const email = await waitForElement(() => getByTestId("email"))
    const name = await waitForElement(() => getByTestId("name"))
    const password = await waitForElement(() => getByTestId("password"))
    const button = await waitForElement(() => getByTestId("submit"))

    email.setAttribute("value", "test@test")
    name.setAttribute("value", "test")
    password.setAttribute("value", "P@55w0rd!")
    button.click()

    await wait(() => expect(onSubmit).toBeCalled())
  })

  it("disables the button while loading", async () => {
    const { getByTestId } = render(<Component loading={true} />)
    const button = await waitForElement(() => getByTestId("submit"))
    expect(button.getAttribute("disabled")).not.toBeNull()
  })
})
