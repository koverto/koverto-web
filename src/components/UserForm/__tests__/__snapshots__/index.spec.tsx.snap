// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm matches the snapshot 1`] = `
<div>
  <form>
    <div>
      <input
        data-testid="name"
        name="name"
        type="text"
      />
    </div>
    <div>
      <input
        data-testid="email"
        name="email"
        type="email"
      />
    </div>
    <div>
      <input
        data-testid="password"
        name="password"
        type="password"
      />
    </div>
    <button
      data-testid="submit"
      type="submit"
    >
      User
    </button>
  </form>
</div>
`;
